packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect } from 'chai';
+import { SnapshotVersion } from '../../../src/core/snapshot_version';
+import { MutationResult } from '../../../src/model/mutation';
+import {
+  PersistentListenStream,
+  PersistentWriteStream,
+  WatchStreamListener,
+  WriteStreamListener
+} from '../../../src/remote/persistent_stream';
+import {
+  DocumentWatchChange,
+  ExistenceFilterChange,
+  WatchTargetChange
+} from '../../../src/remote/watch_change';
+import { AsyncQueue } from '../../../src/util/async_queue';
+import { Deferred } from '../../../src/util/promise';
+import { asyncIt, setMutation } from '../../util/helpers';
+import { withTestDatastore } from '../util/helpers';
+
+/**
+ * StreamEventType combines the events that can be observed by a
+ * WatchStreamListener and WriteStreamListener.
+ */
+type StreamEventType =
+  | 'handshakeComplete'
+  | 'mutationResult'
+  | 'watchChange'
+  | 'open'
+  | 'close';
+
+class StreamStatusListener implements WatchStreamListener, WriteStreamListener {
+  private pendingCallbacks: StreamEventType[] = [];
+  private pendingPromises: Deferred<StreamEventType>[] = [];
+
+  /**
+   * Returns a Promise that resolves when the next callback fires. Resolves the
+   * returned Promise immediately if there is already an unprocessed callback.
+   *
+   * The Promise is rejected if the received callback doesn't match
+   * `expectedCallback`.
+   */
+  awaitCallback(expectedCallback: StreamEventType): Promise<void> {
+    let promise: Promise<StreamEventType>;
+
+    if (this.pendingCallbacks.length > 0) {
+      let pendingCallback = this.pendingCallbacks.shift();
+      promise = Promise.resolve(pendingCallback);
+    } else {
+      const deferred = new Deferred<StreamEventType>();
+      this.pendingPromises.push(deferred);
+      promise = deferred.promise;
+    }
+
+    return promise.then(actualCallback => {
+      if (actualCallback !== expectedCallback) {
+        return Promise.reject(
+          `Unexpected callback encountered. Expected '${expectedCallback}', but got '${actualCallback}'.`
+        );
+      }
```

Comment:
I don't have any strong preference, but FYI I think using expect(...).to.equal(...) here was probably fine.

Reply:
Done.